deploy/helm/kubecf/templates/_config.tpl

```diff
@@ -0,0 +1,290 @@
+|   if all of the AND terms are true.
+|
+| - An AND term is either the string "true" or "false", or it will be evaluated
+|   by looking up it's value in $.kubecf.config.
+|
+| - An AND term may be prefixed by "!" in which case its value is negated.
+|
+| - An AND term may also be a condition expression enclosed in parenthesis.
+|
+| - None of the terms can be an empty string.
+|
+| Example: "!features.foo.enabled && (features.bar.enabled || features.baz.enabled)"
+|
+| Usage examples: This function is used to evaluate the keys of the
+| "deprecation" hash earlier in this file, and the "release.condition" values
+| in assets/operations/releases.yaml.
+==========================================================================================
+*/}}
+{{- define "_config.condition" }}
+  {{- $root := index . 0 }}
+  {{- $condition := index . 1 }}
+
+  {{- if kindIs "invalid" $condition }}
+    {{- /* The absence of a condition (nil) is unconditionally true */}}
+    {{- $condition = true }}
+  {{- end }}
+
+  {{- if kindIs "bool" $condition }}
+    {{- ternary "true" "false" $condition }}
+
+  {{- else }}
+    {{- /* Count the number of left parenthesis to determine the number of groups */}}
+    {{- range $_ := until (sub (len (splitList "(" $condition)) 1 | int) }}
+      {{- /* Find left inner-most group, evaluate it, and replace the expression with the value */}}
+      {{- $group := regexFind "\\([^\\)]+\\)" $condition }}
+      {{- /* There may be fewer groups than left parens if some groups turn out to be identical */}}
+      {{- /* E.g. "((to.be) || !(to.be))" will collapse to "(true || !true)" after the 1st iteration */}}
+      {{- if $group }}
+        {{- $value := include "_config.condition" (list $root ($group | trimPrefix "(" | trimSuffix ")")) }}
+        {{- $condition = replace $group $value $condition }}
+      {{- end }}
+    {{- end }}
+
+    {{- /* Evaluate the remaining expression based on operator precedence: NOT (highest), AND, OR (lowest) */}}
+    {{- $or_value := false }}
+    {{- range $or_term := splitList "||" (nospace $condition) }}
+      {{- $and_value := true }}
+      {{- range $and_term := splitList "&&" $or_term }}
+        {{- $term := trimPrefix "!" $and_term }}
+        {{- /* The term is either literally "true" or "false", or must be looked up */}}
+        {{- $value := true }}
+        {{- if ne $term "true" }}
+          {{- if eq $term "false" }}
+            {{- $value = false }}
+          {{- else }}
+            {{- $value = include "_config.lookup" (list $root $term) }}
+          {{- end }}
+        {{- end }}
+        {{- if hasPrefix "!" $and_term }}
+          {{- $value = not $value }}
+        {{- end }}
+        {{- $and_value = and $and_value $value }}
+      {{- end }}
+      {{- $or_value = or $or_value $and_value }}
+    {{- end }}
+    {{- /* "ternary" requires a real bool for the third argument */}}
+    {{- ternary "false" "true" (not $or_value) }}
```

Comment:
It was inverted because `$or_value` needed to be turned into a true boolean. Maybe I should have used `(not (not $or_value))` and not swap the order of the return values?

I was hoping the comment above this line would explain why this expression looks the way it does.

Reply:
Maybe I should have added: Unlike `and` and `or`, `not` returns a proper boolean, as it can't just return its original argument. So double-negation would be the same as a type cast to boolean.

Reply:
Fixed in d120de6